src/storage/src/source/reclock/compat.rs

```diff
@@ -0,0 +1,448 @@
+                new_custom_frontier
+                    .iter()
+                    .map(|(pid, offset)| ((pid.clone(), *offset), ts, 1)),
+            );
+
+            custom_frontier = new_custom_frontier;
+        }
+        // Then, consolidate the custom updates and we're done
+        consolidation::consolidate_updates(&mut custom_frontier_updates);
+
+        // And finally convert into rows and attempt to append to the shard
+        let mut row_updates = vec![];
+        for ((pid, offset), ts, diff) in custom_frontier_updates {
+            row_updates.push((pack_binding(pid, offset), ts, diff));
+        }
+
+        loop {
+            let updates = row_updates
+                .iter()
+                .map(|(data, time, diff)| ((data, ()), time, diff));
+            let upper = upper.clone();
+            let new_upper = new_upper.clone();
+            match self
+                .write_handle
+                .compare_and_append(updates, upper, new_upper)
+                .await
+            {
+                Ok(Ok(result)) => return result,
+                Ok(Err(invalid_use)) => panic!("compare_and_append failed: {invalid_use}"),
+                // An external error means that the operation might have suceeded or failed but we
+                // don't know. In either case it is safe to retry because:
+                // * If it succeeded, then on retry we'll get an `Upper(_)` error as if some other
+                //   process raced us (but we actually raced ourselves). Since the operator is
+                //   built to handle concurrent instances of itself this safe to do and will
+                //   correctly re-sync its state. Once it resyncs we'll re-enter `mint` and notice
+                //   that there are no updates to add (because we just added them and don't know
+                //   it!) and the reclock operation will proceed normally.
+                // * If it failed, then we'll succeed on retry and proceed normally.
+                Err(external_err) => {
+                    tracing::debug!("compare_and_append failed: {external_err}");
+                    continue;
+                }
+            }
+        }
+    }
+
+    async fn next(
+        &mut self,
+    ) -> Option<(
+        Vec<(Self::FromTime, Self::IntoTime, Diff)>,
+        Antichain<Self::IntoTime>,
+    )> {
+        if !std::mem::replace(&mut self.snapshot_produced, true) {
+            for ((update, _), ts, diff) in self
+                .read_handle
+                .snapshot_and_fetch(self.as_of.clone())
+                .await
+                .expect("local since is not beyond read handle's since")
+            {
+                let binding = unpack_binding(update.expect("invalid row"));
+                self.pending_batch.push((binding, ts, diff));
+            }
+        }
+        while let Some(event) = self.events.next().await {
+            match event {
+                ListenEvent::Progress(new_upper) => {
```

Comment:
Yes, because a snapshot must happen at an `as_of` that is strictly less than the upper (otherwise we wouldn't be able to serve it) and therefore we must see at least one `Progress` message with the upper